blueprints/delivery.service.tls.versions.md

```diff
@@ -0,0 +1,358 @@
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+# Delivery Service TLS Versions
+
+## Problem Description
+Currently, supprted TLS versions cannot be configured. This means that a
+Delivery Service can have clients requesting its content over old, insecure
+versions of TLS, and Tenants controlling that Delivery Service have no way to
+force clients to use more recent versions.
+
+## Proposed Changes
+Changes to ATS in version 8.1 allow cache server configuration generation to
+produce configuration files that can effectively control the TLS versions
+allowed by a Delivery Service, causing clients using other versions to result
+in request failure. Currently, (in unreleased behavior on the master branch),
+this is implemented using a Delivery Service Profile Parameter. However,
+Parameters are notorious for being unvalidated, unstructured data that can be
+just obscure enough to escape notice, even when properly documented. For the
+best possible user, operator, and developer experiences, the data should be
+formally added to Delivery Service objects.
+
+## Data Model Impact
+Delivery Service objects will be augmented to include a set of allowed TLS
+versions for the Delivery Service's content. Specifically, They will be
+expanded to implement this interface:
+
+```typescript
+interface TLSVersionSpecifier {
+	tlsVersions: [string, ...Array<string>] | null;
+}
+```
+
+For those not super familiar with Typescript, this interface specifies a
+`tlsVersions` property that can be either null, or an array of N strings,
+where N is *at least one*. In other words, this can be null, but cannot be an
+empty array.
+
+The reason for this is best understood by first understanding what `null` means
+for this field. If `tlsVersions` is `null`, it means that no TLS version
+restrictions in place; the behavior is the same as today without the field, all
+TLS versions are implicitly supported.
+
+An empty array would seem to signify that no TLS versions are supported, which
```

Comment:
Could we build this behavior into TP by default? Meaning the `TLSVersions` field is set to an empty array by default for `HTTP` delivery services, and is unmodifiable.

Reply:
We can hide the controls, but because all Delivery Service "Types" are still the same "Type" of object we should not overwrite existing tlsVersions. For example, Traffic Router has issues with Delivery Services that don't have SSL keys but do support HTTPS, but you can't generate keys for a DS that doesn't exist. So, you could reasonably create a DS as HTTP-only with everything configured the way you want it, then generate the keys and switch it to use HTTPS

Reply:
Got you, makes sense.

Reply:
I'm not strictly a -1, but UI hacks like this are poor design. It shouldn't be possible for users to assign TLS versions to an HTTP service. If the system can't handle that, we should fix the system.

Reply:
I realize that's not trivial, which is why I'm not -1'ing this. But I do wish we would prioritize fixing those footguns, instead of adding more hacks around them.

Reply:
Breaking DSes out into separate types per Type (or at least routing Type) has been on my to-do list for well over a year now. It's a big ask.